lib/fluent/plugin_helper/storage.rb

```diff
@@ -0,0 +1,306 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'monitor'
+require 'forwardable'
+
+require 'fluent/plugin'
+require 'fluent/plugin/storage'
+require 'fluent/plugin_helper/thread'
+require 'fluent/plugin_helper/timer'
+require 'fluent/config/element'
+
+module Fluent
+  module PluginHelper
+    module Storage
+      include Fluent::PluginHelper::Thread
+      include Fluent::PluginHelper::Timer
+
+      StorageState = Struct.new(:storage, :running)
+
+      def storage_create(usage: '', type: nil, conf: nil)
+        s = @_storages[usage]
+        if s && s.running
+          return s.storage
+        elsif !s
+          unless type
+            raise ArgumentError, "BUG: type not specified without configuration"
+          end
+          storage = Plugin.new_storage(type)
+          storage.owner = self
+          config = if conf && conf.is_a?(Fluent::Config::Element)
+                     conf
+                   elsif conf && conf.is_a?(Hash)
```

Comment:
`conf` is double-checked. `conf.is_a?(Hash)` should be moved to above condition?